pkg/kn/plugin/context_sharing.go

```diff
@@ -0,0 +1,225 @@
+type PluginWithManifest interface {
+	// Plugin original interface wrapper
+	Plugin
+
+	// GetManifest
+	GetManifest() *Manifest
+
+	// GetContextData
+	GetContextData() map[string]string
+
+	// ExecuteWithContext
+	ExecuteWithContext(ctx context.Context, args []string) error
+}
+
+//--TYPES--
+
+var CtxManager *ContextDataManager
+
+type ContextDataManager struct {
+	//ContextData map[string]ContextData `json:"-"`
+	Producers map[string][]string `json:"producers"`
+	Consumers map[string][]string `json:"consumers"`
+	Manifests map[string]Manifest `json:"manifests"`
+}
+
+func NewContextManager() (*ContextDataManager, error) {
+	if CtxManager == nil {
+		//println("opening file...")
+		//file, err := os.Open(filepath.Join(filepath.Dir(config.GlobalConfig.ConfigFile()), "context.json"))
+		//if err != nil {
+		//	return nil, err
+		//}
+		//decoder := json.NewDecoder(file)
+		//ctxManager = &ContextDataManager{}
+		//if err := decoder.Decode(ctxManager); err != nil {
+		//	return nil, err
+		//}
+		//out := new(bytes.Buffer)
+		//enc := json.NewEncoder(out)
+		//enc.SetIndent("", "    ")
+		//enc.Encode(ctxManager)
+		//println(out.String())
+		CtxManager = &ContextDataManager{
+			Producers: map[string][]string{},
+			Consumers: map[string][]string{},
+			Manifests: map[string]Manifest{},
+		}
+	}
+	return CtxManager, nil
+}
+
+// GetConsumesKeys returns array of keys consumed by plugin
+func (c *ContextDataManager) GetConsumesKeys(pluginName string) []string {
+	return c.Manifests[pluginName].ConsumesContextDataKeys
+}
+
+// GetProducesKeys returns array of keys produced by plugin
+func (c *ContextDataManager) GetProducesKeys(pluginName string) []string {
+	return c.Manifests[pluginName].ProducesContextDataKeys
+}
+
+func (c *ContextDataManager) FetchContextData(pluginManager *Manager) context.Context {
+	ctx := context.Background()
+	for _, p := range pluginManager.GetInternalPlugins() {
+		if p.Name() == "kn func" {
+			if pwm, ok := p.(PluginWithManifest); ok {
+				data := pwm.GetContextData()
+
+				for k, v := range data {
+					ctx = context.WithValue(ctx, k, v)
+				}
+			}
+		}
+	}
+	//ctx = context.WithValue(ctx, "service", "hello")
+	return ctx
+
+}
+
+// FetchManifests it tries to retrieve manifest from both inlined and external plugins
+func (c *ContextDataManager) FetchManifests(pluginManager *Manager) error {
+	for _, plugin := range pluginManager.GetInternalPlugins() {
+		manifest := &Manifest{}
+		// For the integrity build the same name format as external plugins
+		pluginName := "kn-" + strings.Join(plugin.CommandParts(), "-")
+		if pwm, ok := plugin.(PluginWithManifest); ok {
+			manifest = pwm.GetManifest()
+		}
+
+		// Add manifest to map
+		c.Manifests[pluginName] = *manifest
+
+		if manifest.HasManifest {
+			c.populateDataKeys(manifest, pluginName)
+		}
+	}
+	plugins, err := pluginManager.ListPlugins()
+	if err != nil {
+		return err
+	}
+	for _, plugin := range plugins {
```

Comment:
I would be also cool, to remove plugins from the cache that are not listed anymore. So checking which keys of the c.Manifests[..] have been touched and remove at the end every entry with a key that is neither an internal or external plugin.